File: client/src/main/java/net/hoz/netapi/client/lang/NetTranslationContainer.java
package net.hoz.netapi.client.lang;

import lombok.*;
import net.hoz.netapi.client.data.DataHolder;
import org.screamingsandals.lib.lang.container.TranslationContainer;
import org.spongepowered.configurate.ConfigurationNode;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

@Data
@RequiredArgsConstructor
public class NetTranslationContainer implements TranslationContainer {
    private final Locale locale;
    private final DataHolder dataHolder;
    @Getter
    @Setter
    private TranslationContainer fallbackContainer;

    @Override
    public ConfigurationNode getNode() {
        return dataHolder.root();
    }

    @Override
    public void setNode(ConfigurationNode configurationNode) {
        //do nothing
    }

    @Override
    public List<String> translate(Collection<String> collection) {
        final var where = String.join(".", collection);
        final var node = dataHolder.node(where);
        if (node.isList()) {
            return node.childrenList()
                    .stream()
                    .map(ConfigurationNode::getString)
                    .collect(Collectors.toList());
        }
        if (!node.empty()) {
            return List.of(node.getString(""));
        }

        return fallbackContainer != null ? fallbackContainer.translate(collection) : List.of();
    }

    @Override
    public List<String> translate(String... strings) {
        return translate(Arrays.asList(strings));
    }

    @Override
    public boolean isEmpty() {
        return dataHolder.root().empty() || (fallbackContainer != null && fallbackContainer.isEmpty());
    }
}

File: client/src/main/kotlin/net/hoz/netapi/client/lang/NetTranslationContainer.kt
package net.hoz.netapi.client.lang

import net.hoz.netapi.client.data.DataHolder
import org.screamingsandals.lib.lang.container.TranslationContainer
import org.spongepowered.configurate.ConfigurationNode
import java.util.*

class NetTranslationContainer(
    val locale: Locale,
    val dataHolder: DataHolder
) : TranslationContainer {
    private var fallbackContainer: TranslationContainer? = null

    override fun getNode(): ConfigurationNode = dataHolder.root

    override fun setNode(configurationNode: ConfigurationNode) {
        // do nothing
    }

    override fun setFallbackContainer(fallbackContainer: TranslationContainer) {
        this.fallbackContainer = fallbackContainer
    }

    override fun getFallbackContainer(): TranslationContainer? = fallbackContainer

    override fun translate(keys: Collection<String>): List<String> {
        val where = keys.joinToString { "." }
        val node = dataHolder.node(where)

        if (node.isList) {
            return node.childrenList().mapNotNull { it.string }
        }

        if (!node.empty()) {
            return listOf(node.getString(""))
        }

        return fallbackContainer?.translate(keys) ?: listOf()
    }

    override fun translate(vararg strings: String): List<String> = translate(strings.toList())

    override fun isEmpty(): Boolean = dataHolder.root.empty() || fallbackContainer?.isEmpty ?: false
}